src/index.c

```diff
@@ -1550,21 +1556,30 @@ int git_index__fill(git_index *index, const git_vector *source_entries)
 
 	assert(index);
 
+	if (!source_entries->length)
+		return 0;
+
 	if (git_mutex_lock(&index->lock) < 0) {
 		giterr_set(GITERR_OS, "Unable to acquire index lock");
 		return -1;
 	}
 
+	if (git_vector_size_hint(&index->entries, source_entries->length) < 0) {
+		git_mutex_unlock(&index->lock);
+		return -1;
+	}
```

Comment:
We should set the size of the hashmap here as well. Resizing that is a lot more expensive than resizing the vector (because of all the hashing in the hashmap).